// @aztec/bb.js/dest/node-cjs/proof/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.flattenFieldsAsArray = exports.deflattenFields = exports.reconstructUltraPlonkProof = exports.reconstructHonkProof = exports.splitHonkProof = exports.AGGREGATION_OBJECT_LENGTH = void 0;
exports.AGGREGATION_OBJECT_LENGTH = 16;
// Fields are 32 bytes
const fieldByteSize = 32;
function splitHonkProof(proofWithPublicInputs, numPublicInputs) {
    const publicInputs = proofWithPublicInputs.slice(0, numPublicInputs * fieldByteSize);
    const proof = proofWithPublicInputs.slice(numPublicInputs * fieldByteSize);
    return {
        proof,
        publicInputs,
    };
}
exports.splitHonkProof = splitHonkProof;
function reconstructHonkProof(publicInputs, proof) {
    const proofWithPublicInputs = Uint8Array.from([...publicInputs, ...proof]);
    return proofWithPublicInputs;
}
exports.reconstructHonkProof = reconstructHonkProof;
function reconstructUltraPlonkProof(proofData) {
    // Flatten publicInputs
    const publicInputsConcatenated = flattenFieldsAsArray(proofData.publicInputs);
    // Concatenate publicInputs and proof
    const proofWithPublicInputs = Uint8Array.from([...publicInputsConcatenated, ...proofData.proof]);
    return proofWithPublicInputs;
}
exports.reconstructUltraPlonkProof = reconstructUltraPlonkProof;
function deflattenFields(flattenedFields) {
    const publicInputSize = 32;
    const chunkedFlattenedPublicInputs = [];
    for (let i = 0; i < flattenedFields.length; i += publicInputSize) {
        const publicInput = flattenedFields.slice(i, i + publicInputSize);
        chunkedFlattenedPublicInputs.push(publicInput);
    }
    return chunkedFlattenedPublicInputs.map(uint8ArrayToHex);
}
exports.deflattenFields = deflattenFields;
function flattenFieldsAsArray(fields) {
    const flattenedPublicInputs = fields.map(hexToUint8Array);
    return flattenUint8Arrays(flattenedPublicInputs);
}
exports.flattenFieldsAsArray = flattenFieldsAsArray;
function flattenUint8Arrays(arrays) {
    const totalLength = arrays.reduce((acc, val) => acc + val.length, 0);
    const result = new Uint8Array(totalLength);
    let offset = 0;
    for (const arr of arrays) {
        result.set(arr, offset);
        offset += arr.length;
    }
    return result;
}
function uint8ArrayToHex(buffer) {
    const hex = [];
    buffer.forEach(function (i) {
        let h = i.toString(16);
        if (h.length % 2) {
            h = '0' + h;
        }
        hex.push(h);
    });
    return '0x' + hex.join('');
}
function hexToUint8Array(hex) {
    const sanitisedHex = BigInt(hex).toString(16).padStart(64, '0');
    const len = sanitisedHex.length / 2;
    const u8 = new Uint8Array(len);
    let i = 0;
    let j = 0;
    while (i < len) {
        u8[i] = parseInt(sanitisedHex.slice(j, j + 2), 16);
        i += 1;
        j += 2;
    }
    return u8;
}
